api/envoy/config/endpoint/v3/load_report.proto

```diff
@@ -136,12 +138,24 @@ message ClusterStats {
   }
 
   // The name of the cluster.
-  string cluster_name = 1 [(validate.rules).string = {min_len: 1}];
+  string cluster_name = 1 [
```

Comment:
One thing I've been thinking about is whether we are better off just using flattened string UDPA URLs everywhere other than DiscoveryRequest/Response. Within Envoy, I anticipate that we will flatten to string these URLs rather than maintain a structured representation, since this makes it easy to interop with existing opaque string names.

Reply:
I think there are two independent places we can make this choice:

1. In the client implementation.
2. In the API.

The two choices are largely orthogonal, because regardless of what choice we make for either of them, we can easily convert to the other when we send or receive messages.

The decision for (1) does not have to be universal -- Envoy and gRPC (and other clients in the future) could all make different choices for their internal representations.  For gRPC C-core, I am thinking that we will probably use a struct representation, because there are a lot of places where we're going to want to access just one of the fields from the URI (e.g., just the authority), and we don't want to have to re-parse the URI each time we need to do that.  But I might change my mind about that as I see how the code winds up looking.

In any case, though, this PR is only about the API.  I think we should be consistent about the representation across the entire API.  If we're using the UDPA protos in DiscoveryRequest/Response, then I think we should also use them everywhere else in the API as well.

Reply:
I buy the consistency argument, but I'd also argue there is a distinction between what we opt for in the transport protocol and what we use in the resources themselves. There's probably other places we might have missed, e.g. what about `WeightedCluster` or every place we reference SDS?

Reply:
Good catch!  I've added the needed field in `WeightedCluster`.  I also added one for VHDS.

It looks like you already took care of SDS in #11810 (see [here](https://github.com/envoyproxy/envoy/blob/1ca53123c078b6545d393d2974c58ddc3bebabb4/api/envoy/extensions/transport_sockets/tls/v3/secret.proto#L44)), at the same time that you did RDS and EDS.

Do we also need something for SRDS?  I don't know enough about how that works to know what new fields might needed or where they should go.

Reply:
Try grep or search for `string cluster` (which is just one set of these names). There's actually quite a lot, ranging from places such as `cluster_header` (do we force our naming scheme on that?), `http_uri.proto` (basically, anytime we try and do HTTP in Envoy we need to know a cluster name, etc.)

Reply:
I've added a UdpaResourceLocator field everywhere that I see "string cluster", with only one exception.  That exception is in [config_source.proto](https://github.com/envoyproxy/envoy/blob/8281dd61867bfa7cf67c54389190c29ecac8d76d/api/envoy/config/core/v3/config_source.proto#L91), where the cluster name says that it's intended to be used only with REST and only when the specified cluster is fully static (no CDS or EDS).  I don't think we'll wind up using the new names with those constraints, so I didn't add the field there.

Reply:
Do we also need to do this for secret name? After going through this exercise, are you still preferring this approach?

Reply:
Which "secret name" field are you referring to?  I think you already took care of SDS, as I mentioned above, so I'm not sure what else needs to change here.

With regard to the broader question, I can see arguments both for doing this and for just using simple string representation.  But as I said earlier, I think we should be consistent in how we do this across the entire API.  So I think we have two options:

1. Use string form everywhere.  We would completely remove the `UdpaResourceName` and `UdpaResourceLocator` protos, since they would never be used.  We can simply store URIs in string form in the existing string fields for resource names, both in `{Delta,}Discovery{Request,Response}` and in the resource protos themselves.

2. Go forward with this PR so that we're consistently using `UdpaResourceName` and `UdpaResourceLocator` everywhere.

Maybe we should sit down and game through all of the implications of both options.  For example, in option 1, we'd basically be giving up proto validation, and we'd be requiring all xDS clients and servers to parse and validate the URI strings themselves.  We should also talk through which option will be easier in terms of migration.  Are there going to be problems if both old and new names are encoded in the same string field?  Or can we seamlessly define a bidirectional lossless conversion between old and new names (e.g., the old name just gets moved into the id field somehow)?

Let's talk this through when you get back.